File: Machine Learning/1.1.2 Data visualization/ML.R
library("ROCR")

#读取数据
data_mat <- read.csv("/Users/xinzhe/Desktop/Lulab/毕设/gitbook/BreastCancer.csv", header = T)
for(i in 1:length(data_mat[,7])){
  if(is.na(data_mat[i,7]))
    data_mat[i,7] <- 0
}
rm(i)

#数据归一化
X <- rep(0,9*699)
dim(X) <- c(699, 9)
for(i in 2:(dim(data_mat)[2]-1)){
  X[,i-1] <- (data_mat[, i] - min(data_mat[, i])) / (max(data_mat[, i])-min(data_mat[, i]))
}

#PCA
X_pca <- prcomp(X, center = T, scale = F, rank. = 2)

#PCA可视化
colors <- rainbow(length(unique(data_mat[,11])))
colors_plot <- as.character(data_mat[,11])
colors_plot[which(colors_plot == "benign")] <- colors[1]
colors_plot[which(colors_plot == "malignant")] <- colors[2]
plot(X_pca$x, col = colors_plot, asp = 1, pch = 20,
     xlab = "component_1", ylab = "component_2", main = "PCA plot")
legend("topright", c("benign", "malignant"), col=c(colors[1], colors[2]), lty = 1)

#tSNE
X_tsne <- Rtsne(X, dims = 2, check_duplicates = F)

#tSNE可视化
colors <- rainbow(length(unique(data_mat[,11])))
colors_plot <- as.character(data_mat[,11])
colors_plot[which(colors_plot == "benign")] <- colors[1]
colors_plot[which(colors_plot == "malignant")] <- colors[2]
plot(X_tsne$Y, col = colors_plot, asp = 1, pch = 20,
     xlab = "tSNE_1", ylab = "tSNE_2", main = "tSNE plot")
legend("topleft", c("benign", "malignant"), col=c(colors[1], colors[2]), lty = 1)

File: Machine Learning/1.1.2 Data visualization/ML.py
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.preprocessing import MinMaxScaler
import matplotlib.pyplot as plt
import seaborn as sns

#读取数据
data_mat = []
label_mat = []
with open('/Users/xinzhe/Desktop/Lulab/毕设/gitbook/BreastCancer.csv', 'r') as f:
    for line in f.readlines()[1:]:
        line_arr = line.strip().split(',')
        if line_arr[6] == 'NA':
            line_arr[6] = 0
        data_mat.append([float(line_arr[1]), float(line_arr[2]), float(line_arr[3]), float(line_arr[4]),
        float(line_arr[5]), float(line_arr[6]), float(line_arr[7]), float(line_arr[8]), float(line_arr[9])])
        label_mat.append(line_arr[10])

#数据归一化
scaler = MinMaxScaler()
X = scaler.fit_transform(data_mat)

#PCA
transform = PCA(2)
X_pca = transform.fit_transform(X)
X_, Y_ = X_pca, label_mat

#PCA可视化
pca_data = pd.DataFrame(X_)
pca_data.columns = ['PCA_1', 'PCA_2']
pca_data['label'] = np.array(Y_)
g = sns.scatterplot(x = "PCA_1", y = "PCA_2", hue = "label", data = pca_data, s = 50)
plt.legend(loc = 'best')
plt.show()
plt.close()

#tSNE
transform = TSNE(2)
X_tsne = transform.fit_transform(X)
X_, Y_ = X_tsne, label_mat

##tSNE可视化
tSNE_data = pd.DataFrame(X_)
tSNE_data.columns = ['tSNE_1', 'tSNE_2']
tSNE_data['label'] = np.array(Y_)
g = sns.scatterplot(x = "tSNE_1", y = "tSNE_2", hue = "label", data = tSNE_data, s = 50)
plt.legend(loc = 'best')
plt.show()
plt.close()
